core/object/class_db.cpp

```diff
@@ -545,6 +545,15 @@ Object *ClassDB::instantiate(const StringName &p_class) {
 	return ti->creation_func();
 }
 
+Object *ClassDB::construct_object(Object *(*p_create_func)(), ObjectNativeExtension *p_extension) {
+	if (p_extension) {
+		initializing_with_extension = true;
+		initializing_extension = p_extension;
+		initializing_extension_instance = p_extension->create_instance(p_extension->class_userdata);
+	}
+	return p_create_func();
```

Comment:
should we have an else here that clears the data? can we trust the data currently held in p_extension is trustworthy? I'm probably just being paranoid :)

Reply:
Well, the pointers doesn't really get cleared, but the boolean gets reset inside the object creation function (`p_create_func`), when `instance_get_native_extension_data` is called like it used before:
https://github.com/godotengine/godot/blob/f9ce9a8e103b4b53ad3de032220c3e8799d52409/core/object/class_db.cpp#L508-L518

I didn't want to change the ClassDB logic too much, but I guess we can even try to completely get rid of the static variables.

> can we trust the data currently held in p_extension is trustworthy?

No, and neither is the constructor, but I wouldn't worry too much for now.
Extensions are loaded as dynamic libraries, and will share memory with Godot, so they can do all sort of nasty things anyway unless we implement a sandbox at some point (or some form or separate runtime like embedding wasm).